src/sparseml/transformers/finetune/text_generation.py

```diff
@@ -0,0 +1,343 @@
+
+    # Detecting last checkpoint.
+    last_checkpoint = None
+    if (
+        os.path.isdir(training_args.output_dir)
+        and training_args.do_train
+        and not training_args.overwrite_output_dir
+    ):
+        last_checkpoint = get_last_checkpoint(training_args.output_dir)
+        if last_checkpoint is None and (len(os.listdir(training_args.output_dir)) > 0):
+            raise ValueError(
+                f"Output directory ({training_args.output_dir}) already "
+                "exists and is not empty. Use --overwrite_output_dir to overcome."
+            )
+        elif (
+            last_checkpoint is not None and training_args.resume_from_checkpoint is None
+        ):
+            _LOGGER.info(
+                f"Checkpoint detected, resuming training at {last_checkpoint}. To "
+                "avoid this behavior, change  the `--output_dir` or add "
+                "`--overwrite_output_dir` to train from scratch."
+            )
+
+    # Set seed before initializing model.
+    set_seed(training_args.seed)
+
+    # Load pretrained model
+    # The .from_pretrained methods guarantee that only one local process can
+    # concurrently download model & vocab.
+    config = AutoConfig.from_pretrained(
+        model_args.config_name
+        if model_args.config_name
+        else model_args.model_name_or_path,
+        cache_dir=model_args.cache_dir,
+        revision=model_args.model_revision,
+        use_auth_token=True if model_args.use_auth_token else None,
+    )
+
+    model_kwargs = {
+        "config": config,
+        "cache_dir": model_args.cache_dir,
+        "revision": model_args.model_revision,
+        "use_auth_token": True if model_args.use_auth_token else None,
+    }
+    # this calls from_pretrained under the hood so should be FSDP safe
+    model = SparseAutoModel.text_generation_from_pretrained(
+        model_name_or_path=model_args.model_name_or_path,
+        model_type="model",
+        **model_kwargs,
+    )
+
+    # initialize structure of input model from recipe if needed
+    model_path = model_args.model_name_or_path
+    recipe_path = os.path.join(model_path, "recipe.yaml")
+    if not os.path.exists(recipe_path):
+        _LOGGER.warning(f"No recipes were applied for {model_path}.")
+    else:
+        _LOGGER.warning(f"Applying recipe {recipe_path} to {model_path}")
+        apply_recipe_structure_to_model(model, recipe_path, model_path)
+
+    # Load tokenizer
+    tokenizer_src = (
+        model_args.tokenizer_name
+        if model_args.tokenizer_name
+        else get_shared_tokenizer_src(model, None)
+    )
+    tokenizer = AutoTokenizer.from_pretrained(
+        tokenizer_src,
+        cache_dir=model_args.cache_dir,
+        use_fast=True,
+        revision=model_args.model_revision,
+        use_auth_token=True if model_args.use_auth_token else None,
+    )
+
+    # Load datasets
+    # TODO: will any of this cause problems with FSDP?
+    splits = data_args.splits
```

Comment:
would also be good to move this to a helper fn unless it requires too many loose vars to be passed around